### ISSUE TITLE
[RFC] Use unary left operators to expand variable lookup.

### ISSUE THREAD

**Issue by mhermier** [wren-lang/wren#955]
Hi,
Currently I think there are annoying issues in variable lookup:
- It is not possible for a method to lookup for a variable in a parent scope.
- Symbols that start capitalized letters are the only symbols to be looked in the top-level module scope.
This as various consequence in how you express your code, and might have some more if UTF-8 naming become a reality.

I would like to add to operators to fix those respectively. An example code, using '..' and '...' (but can be anything else) of what I think it can allow :
```
var a = 1
{
  var a = 2
  class Foo {
    a { 3 }
    call() {
      System.print(a)    // 3
      System.print(..a)  // 2
      System.print(...a) // 1
    }
  }
}

```
Incidentally if `..` and `...` use used than `.` could also be used as a short-hand for `this.` and make the syntax a little bit more coherent ?

Comments welcome.

**Comment by PureFox48** [comment]
I think you may have something there though I'd prefer`::` and `:::` to be the operators which would be more in keeping with scope resolution operators in other languages such as C++

**Comment by mhermier** [comment]
I also thought about `::` but because of the way we access static methods I thought it would be more natural to use a variation of `.`, but I'm not final on the decision.
Using `.` as a short hand for `this.` is also meant to ease writing of Capital letters variables, like enum/constant getters.

**Comment by PureFox48** [comment]
TBH, I'm not so keen on using `.` as a shorthand for `this.`.

I like there to be just one way of doing things so I don't have to decide which one to use each time, but that's just me :)

**Comment by mhermier** [comment]
Agreed, but it is the only simple way I found to bypass the uppercase limitation, without getting too much in the way. It is optional to  the RFC, but I think it should be considered if `..` and `...` are envisioned as a revision/improvement of the lookup.

**Comment by PureFox48** [comment]
Fair enough.

Let's wait and see what others think to these ideas as a whole.

**Comment by ruby0x1** [comment]
```js
var a = 1
{
  var a = 2
  class Foo {
    a { 3 }
    call() {
      System.print(a)    // 3
      System.print(..a)  // 2
      System.print(...a) // 1
      class Nested {
        call() { ...a }  // ?
      }
    }
  }
}
```

**Comment by PureFox48** [comment]
I'd say the answer to that would still be `1` as there are no more `a`'s in the Nested class.

**Comment by mhermier** [comment]
```javascript
var a = 1
{
  var a = 2
  class Foo {
    a { 3 }
    call() {
      System.print(a)    // 3
      System.print(..a)  // 2
      System.print(...a) // 1
      var proxy = this
      class Nested {
        call() {
          System.print(a) // error
          System.print(..proxy.a) // 3
          System.print(..a)  // 2 (looks at the first shadowed value below current)
          System.print(...a) // 1 (lookup start from the base of the module, no mater how many nest context there are, it bypass shadowing)
      }
    }
  }
}
```

**Comment by mhermier** [comment]
But this is a concern if we fix multiple invocation of nested class first (unless that has already been fixed).

**Comment by ChayimFriedman2** [comment]
@ruby0x1, classes in Wren don't close over variables, so this is not something worth worrying.

I'm afraid `..` and `...` collide (mentally) with the range operators, and we may also want infinite ranges in the future.

**Comment by ruby0x1** [comment]
Language features that are contextually unclear or inconsistent are most definitely worth worrying about.

**Comment by mhermier** [comment]
Another combination that is visually representative would be `^` for parent lookup (as in lookup up the tree), and `.` for global lookup (as in lookup from nothing/root). But have mixed feelings because last one would really feat to simplify writing as short-hand...
Using `::` does not seems consistent when writing `::Foo.bar(baz)`, unless (crazy idea ahead) `::` was the real lookup operator and make `.` could be user defined ??

**Comment by mhermier** [comment]
To elaborate more on `^` I have found an interesting pro that is pretty unique: It can be multiplied to specify from what level the lookup starts. Variable shadowing is a bad practice, but in case there is non avoidable naming clash, or when coding before cleaning it can be an interesting feature I guess. In practice it would look like:
```javascript
var a = 1
{
  {
    var a = 2
    {
      System.print(^a) // 2
      System.print(^^a) // 1
      System.print(^^^a) // 1
      System.print(^^^^a) // 1
    }
  }
}
```

**Comment by PureFox48** [comment]
I wonder if it would suffice in practice for us just to have a way of specifying that something was declared at module level? Say prefixing the identifier with `^` or `::`.

We could then access module level variables and functions from within a class even if they began with lower case letters or were shadowed.

Of course, this wouldn't deal with nested situations but how common are those anyway and one can always using a proxy if it's necessary to access shadowed or lower case variables at a higher scope level .

**Comment by mhermier** [comment]
Biggest problem is that module level is common and shared for every module,
meaning we have a limited pool of available entries. So even if proxying is
definitively a solution, it has a hard limit and it's implications in terms
of memory cache access. So I think it should be important to think to
solutions to limit it usage before it generalize.

**Comment by PureFox48** [comment]
Yeah, I know it wouldn't be perfect but it would be easy to understand and have a major impact on the 'lower case letter' problem.

There are other languages which distinguish syntactically between global and local scope (notably Python) but I can't think of any with greater granularity than that.

**Comment by mhermier** [comment]
Well if `^` is adopted, it would be a little cumbersome to use in some
place, but at least it would solve the 2 problems at once, at the potential
cost of repeating `^`. Considering shadowing and excecive nesting is a bad
practice,  it should be pretty uncommon to require more than one `^` in 99%
of the cases. That would solve both problem at once, until we really need a
reason to introduce a global lookup operator.

**Comment by mhermier** [comment]
Basic code is in place and is somehow functional. But I have to fix lookup
since it doesn't work as I thought. So I need to think on plan to solve the
issue...

**Comment by clsource** [comment]
I like the ideas but I feel that overusing `^` feels odd.

Maybe another way of implementing this would be similar to a dictionary

- `^a` would look up 1
- `^^a` would look global 
- `^{2:a}` would look up twice (n)

**Comment by mhermier** [comment]
If number is the way no need to make things over-complicated, `^2a` is not ambiguous. Though I would prefer if it does stay a short hand writing. Global lookup in wren is a little bit special since it implies forward declaration, so I think it should really be separated from lookup.
